From 25b33ecb6ca47b80c6f828ba9d08ad1b2f0b0925 Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Mon, 10 Apr 2023 16:38:30 +0800
Subject: [PATCH 0887/1646] [Huawei] userswap: provide cpu info in userfault
 msg

Offering: HULK
hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

The uffd_msg.reserved3 field is used to transfer the CPU information of
the PF.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 fs/userfaultfd.c         | 3 +++
 include/linux/userswap.h | 6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/fs/userfaultfd.c b/fs/userfaultfd.c
index 8a916c96d967..d27b66eb5aa0 100644
--- a/fs/userfaultfd.c
+++ b/fs/userfaultfd.c
@@ -218,6 +218,9 @@ static inline struct uffd_msg userfault_msg(unsigned long address,
 		msg.arg.pagefault.flags |= UFFD_PAGEFAULT_FLAG_WP;
 	if (features & UFFD_FEATURE_THREAD_ID)
 		msg.arg.pagefault.feat.ptid = task_pid_vnr(current);
+#ifdef CONFIG_USERSWAP
+	uswap_get_cpu_id(reason, &msg);
+#endif
 	return msg;
 }
 
diff --git a/include/linux/userswap.h b/include/linux/userswap.h
index 29a3bead95d0..9d1d0ee2e9d7 100644
--- a/include/linux/userswap.h
+++ b/include/linux/userswap.h
@@ -66,6 +66,12 @@ static inline bool uswap_missing(struct vm_area_struct *vma)
 	return false;
 }
 
+static inline void uswap_get_cpu_id(unsigned long reason, struct uffd_msg *msg)
+{
+	if (reason & VM_USWAP)
+		msg->reserved3 = smp_processor_id();
+}
+
 #endif /* CONFIG_USERSWAP */
 
 #endif /* _LINUX_USERSWAP_H */
-- 
2.17.1

